controller/engine_controller.go

```diff
@@ -1814,6 +1828,67 @@ func (ec *EngineController) startRebuilding(e *longhorn.Engine, replica, addr st
 	return nil
 }
 
+// updateReplicaRebuildFailedCondition updates the rebuild failed condition if replica rebuilding failed
+func updateReplicaRebuildFailedCondition(log *logrus.Entry, ds *datastore.DataStore, replicaName, errMsg string) error {
+	replica, err := ds.GetReplica(replicaName)
+	if err != nil {
+		return err
+	}
+
+	replicaRebuildFailedReason, conditionStatus, err := getReplicaRebuildFailedReason(ds, replica.Spec.NodeID, errMsg)
+	if err != nil {
+		return err
+	}
+
+	replica.Status.Conditions = types.SetCondition(
+		replica.Status.Conditions, longhorn.ReplicaConditionTypeRebuildFailed,
+		conditionStatus, replicaRebuildFailedReason, errMsg)
+
+	_, err = ds.UpdateReplicaStatus(replica)
+
+	return err
+}
+
+func getReplicaRebuildFailedReason(ds *datastore.DataStore, replicaNodeID, errMsg string) (string, longhorn.ConditionStatus, error) {
+	replicaRebuildFailedReason, conditionStatus, isRebuildingFailedByNetwork := getReplicaRebuildFailedReasonFromError(errMsg)
+	if isRebuildingFailedByNetwork {
+		replicaKubernetesNode, err := ds.GetKubernetesNode(replicaNodeID)
+		if err != nil {
+			return "", "", err
+		}
+		for _, con := range replicaKubernetesNode.Status.Conditions {
+			if con.Type != v1.NodeReady {
+				continue
+			}
+			if con.Status != v1.ConditionTrue {
+				replicaRebuildFailedReason = longhorn.ReplicaConditionReasonRebuildFailedNodeDown
+			}
+			break
+		}
+	}
+
+	return replicaRebuildFailedReason, conditionStatus, nil
+}
+
+func getReplicaRebuildFailedReasonFromError(errMsg string) (string, longhorn.ConditionStatus, bool) {
+	switch {
+	case strings.Contains(errMsg, longhorn.ReplicaRebuildFailedDisconnectionErrorMSG):
+		return longhorn.ReplicaConditionReasonRebuildFailedDisconnection, longhorn.ConditionStatusTrue, true
+	case strings.Contains(errMsg, longhorn.ReplicaRebuildFailedConnectionResetErrorMSG):
+		return longhorn.ReplicaConditionReasonRebuildFailedConnectionReset, longhorn.ConditionStatusTrue, true
+	case strings.Contains(errMsg, longhorn.ReplicaRebuildFailedConnectionTimeoutErrorMSG):
+		return longhorn.ReplicaConditionReasonRebuildFailedConnectionTimeout, longhorn.ConditionStatusTrue, true
+	case strings.Contains(errMsg, longhorn.ReplicaRebuildFailedConnectionRefusedErrorMSG):
+		return longhorn.ReplicaConditionReasonRebuildFailedConnectionRefused, longhorn.ConditionStatusTrue, true
+	case strings.Contains(errMsg, longhorn.ReplicaRebuildFailedReadingServerEOFErrorMSG):
+		return longhorn.ReplicaConditionReasonRebuildFailedReadingServerEOF, longhorn.ConditionStatusTrue, true
+	case errMsg == "":
+		return "", longhorn.ConditionStatusFalse, false
+	default:
+		return longhorn.ReplicaConditionReasonRebuildFailedGeneral, longhorn.ConditionStatusTrue, false
+	}
```

Comment:
Because the error you handled is from a gRPC call, so basically you should be able to check gRPC error code to determine what kind of error it is. 

Suggest using `status.FromError` to get the converted gRPC error then check its error code. If the code is `Unavailable`, `Canceled` or `DeadlineExceeded`, then we could guess that's caused by unstable network. You don't need to hard code message then do regex or str check to determine the error type anymore. 

If you would like to cover the node condition from node status to superseded the gRPC error code, and it should be fine. However, I would say, let's just focus on the gRPC code which will make the thing clear.

Reply:
ref: backing image manager used that as well for different code

[/pkg/util/util.go#L81-L84](https://github.com/longhorn/backing-image-manager/blob/3cb922d175822c1b9b22fc0573abf6a899468648/pkg/util/util.go#L81-L84)
```go
func IsGRPCErrorMatchingCode(err error, errCode codes.Code) bool {
	gRPCStatus, ok := status.FromError(err)
	return ok && gRPCStatus.Code() == errCode
}
```

Reply:
I tried to get the gRPC status by `status.FromError`, it seems that we have two layers of the gRPC call from `instance-manager` to `engine`. The message would be like this:
```txt
      'proxyServer=10.42.3.40:8501 destination=10.42.3.40:10000: failed to
      add replica tcp://10.42.1.38:10000 for volume: rpc error: code = Unknown desc
      = failed to sync files [{FromFileName:volume-snap-789558f4-89af-42d6-8e1d-f615ce067d14.img
      ToFileName:volume-snap-789558f4-89af-42d6-8e1d-f615ce067d14.img ActualSize:0}
      {FromFileName:volume-snap-789558f4-89af-42d6-8e1d-f615ce067d14.img.meta ToFileName:volume-snap-789558f4-89af-42d6-8e1d-f615ce067d14.img.meta
      ActualSize:0} {FromFileName:volume-snap-9db919f3-1737-410b-84f5-00743278ee5c.img
      ToFileName:volume-snap-9db919f3-1737-410b-84f5-00743278ee5c.img ActualSize:10737442816}
      {FromFileName:volume-snap-9db919f3-1737-410b-84f5-00743278ee5c.img.meta ToFileName:volume-snap-9db919f3-1737-410b-84f5-00743278ee5c.img.meta
      ActualSize:0}] from tcp://10.42.4.41:10000: rpc error: code = Unavailable desc
      = error reading from server: read tcp 10.42.3.40:47520->10.42.1.38:10002: read:
      connection reset by peer'
```
I can get the first gRPC status from errors `rpc error: code = Unknown desc = failed to sync files ...` returned by instance-manager but it might be always `Unknown` and it seems that second error `rpc error: code = Unavailable desc = error reading from server: read tcp 10.42.3.40:47520->10.42.1.38:10002: read: connection reset by peer'` is wrapped into the message of first error and it can not be transformed back to the gRPC status (I don't have any idea).
So I scale down the number of errMsg to `code = Canceled`, `code = DeadlineExceeded` and `code = Unavailable` to check if the error is caused by network.

Reply:
Exactly, this is not what I thought, because I assumed the error should be just returned back w/o any modification because the call is from the proxy. However, it's actually modified in the replica client which is used by the proxy server.

[/vendor/github.com/longhorn/longhorn-engine/pkg/replica/client/client.go#L503-L511](https://github.com/longhorn/longhorn-instance-manager/blob/cb92c147d5f8deb55cbe776d7d6f42e1f22ae2f3/vendor/github.com/longhorn/longhorn-engine/pkg/replica/client/client.go#L503-L511)
```go
	if _, err := syncAgentServiceClient.FilesSync(ctx, &ptypes.FilesSyncRequest{
		FromAddress:               fromAddress,
		ToHost:                    c.host,
		SyncFileInfoList:          syncFileInfoListToSyncAgentGRPCFormat(list),
		FastSync:                  fastSync,
		FileSyncHttpClientTimeout: int32(fileSyncHTTPClientTimeout),
	}); err != nil {
		return errors.Wrapf(err, "failed to sync files %+v from %v", list, fromAddress)
	}
```